### PR TITLE
OSD-12042 add a script to replace master

### PR DESCRIPTION
**Pull request by feichashao**
Add a script to help replacing the master machine. Input the machine name of the master, and it will delete that master machine and create a new one with same configuration.

### CODE REVIEW

File: scripts/SREP/replace-master/script.sh [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                  --argjson spec "$SPEC" \
+                  '{ "apiVersion": $apiVersion, "kind": "Machine", "metadata": { "name": $name, "namespace": $namespace, labels: $labels}, "spec": $spec }' )
+
+### delete the master machine
+### usually the oc delete command will block until the machine is gone.
+### in some cases when the master node is gone, the oc command will fail to watch the status, so sleep 300s in such case.
+oc delete machine "$machine" -n $MACHINE_NS || sleep 300
+
+### check if only 2 running masters are left
+echo "INFO: wait for the original master gone"
```
Comment (supreeth7) [inline scripts/SREP/replace-master/script.sh:108]:
```suggestion
echo "INFO: waiting for the original master to be deleted..."
```

File: scripts/SREP/replace-master/script.sh [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    then
+       echo "ERROR: etcd $node not in started state"
+       return 1
+    fi
+done
+}
+
+replace_machine() {
+echo "INFO: replacing the master machine"
+ORIGN_MASTER_JSON=$(oc -n "$MACHINE_NS" get machine "$machine" -o json)
```
Comment (supreeth7) [inline scripts/SREP/replace-master/script.sh:89]:
```suggestion
OLD_MASTER_JSON=$(oc -n "$MACHINE_NS" get machine "$machine" -o json)
```

File: scripts/SREP/replace-master/script.sh [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+### usually the oc delete command will block until the machine is gone.
+### in some cases when the master node is gone, the oc command will fail to watch the status, so sleep 300s in such case.
+oc delete machine "$machine" -n $MACHINE_NS || sleep 300
+
+### check if only 2 running masters are left
+echo "INFO: wait for the original master gone"
+sleep 600 # wait some time here for etcd election.
+if [[ $(oc get node -l node-role.kubernetes.io/master --no-headers | wc -l) -ne 2 ]]
+then
+    echo "ERROR: The living master nodes not equal 2 after deleting $machine"
```
Comment (supreeth7) [inline scripts/SREP/replace-master/script.sh:112]:
```suggestion
    echo "ERROR: Current master nodes are not equal to 2 after deleting $machine"
```

File: scripts/SREP/replace-master/script.sh [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    return 1
+fi
+
+echo "INFO: creating the new master"
+oc create -f - <<< "$NEW_MASTER_JSON"
+
+### wait until all 3 masters are running
+echo "INFO: wait for the new master become ready"
```
Comment (supreeth7) [inline scripts/SREP/replace-master/script.sh:120]:
```suggestion
echo "INFO: waiting for the new master to become ready..."
```

File: scripts/SREP/replace-master/script.sh [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+sleep 600 # it takes at most 10 mins for a machine to become ready otherwise machine-api will delete and retry.
+if [[ $( oc get node -l node-role.kubernetes.io/master --no-headers | { grep -iv "NotReady" || true; } | wc -l) -ne 3 ]]
+then
+    echo "ERROR: Ready master nodes not equal 3"
```
Comment (supreeth7) [inline scripts/SREP/replace-master/script.sh:124]:
```suggestion
    echo "ERROR: Master nodes with the state 'Ready' are not equal to 3"
```

File: scripts/SREP/replace-master/script.sh [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+ETCD_MEMBERS=$(oc rsh -n $ETCD_NS -c etcdctl "$ETCD_POD" etcdctl member list -w simple)
+ETCD_MEM_TO_DELETE=$ETCD_MEMBERS
+for master in $LIVING_MASTERS
+do
+    ETCD_MEM_TO_DELETE=$({ grep -v "$master" || true; } <<< "$ETCD_MEM_TO_DELETE")
+done
+
+if [[ $(wc -l <<< "$ETCD_MEM_TO_DELETE") -gt 1 ]]
+then
+    echo "ERROR: There are more than 1 etcd members have no corresponding running masters, need human investigation"
```
Comment (supreeth7) [inline scripts/SREP/replace-master/script.sh:158]:
```suggestion
    echo "ERROR: There are more than 1 etcd members having no corresponding running masters, further human investigation needed"
```

File: scripts/SREP/replace-master/script.sh [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    if [[ $( { grep "true" || true; } <<< "$EP_HEALTH_INFO" | wc -l ) -eq 3 ]]
+    then
+        SUCCEED=1
+        break
+    fi
+done
+
+if [[ $SUCCEED -eq 1 ]]
+then
+    echo "ETCD has 3 healthy members, master replacement succeed"
```
Comment (supreeth7) [inline scripts/SREP/replace-master/script.sh:186]:
```suggestion
    echo "ETCD has 3 healthy members, master replacement successful!"
```

**Review by supreeth7 — CHANGES_REQUESTED** [review]
Thank you for working on this, this script would be super helpful! 
Took a quick pass through the code, didn't get a chance to test it though, found a few nits.

Suggestion: I've noticed `validate_input()` doing a lot more than what its name suggests, perhaps we can fragment the code to an extent for better readability. I would prefer breaking it down into additional modular functions like `checkClusterVersion()`,`isUpgrading()`,`validateMachine()`,`checkMastersHealth()`,`checkETCDHealth()`. Modularity for the win 🏆

**Comment by feichashao** [comment]
@supreeth7  Thank you very much for reviewing and providing suggestions!  I changed the script to follow the coding style and put the checks to `preflight_check()`.  I didn't further modulized that because each check is a few lines of code and need to reuse some info from the previous check, putting them into dedicated functions will become another wrapper of the oc command and need passing info around. That's said, I am still a fan of modularity :D

**Comment by supreeth7** [comment]
Thanks @feichashao, looks great! 
Happy to merge once confirmed this has been tested in stage.

**Comment by MitaliBhalla** [comment]
> Thanks @feichashao, looks great! Happy to merge once confirmed this has been tested in stage.

I have tested this on stage for 4.8,4.9 and 4.10 clusters. Good to go from my side.

**Pull request merged**